### PR TITLE
Fix S2589 FN: Infinite for-loop

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Follow-up to #7795 
Now that we visit loop conditions up to three times, one branch usually leaves the loop on the latest visit. This makes the arbitrary branching we have done before unnecessary. Removing it allows us to find infinite for-loops:

```
public void InfiniteForLoop(string arg)
{
    for (var i = 0; i < 10; i--)    // FN
    {
    }
}
```

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.CSharp8.cs [hunk 1]
```diff
@@ -392,17 +392,17 @@ void Method()
...
+            if (sb is null)             // Noncompliant FP
             {
                 sb = new StringBuilder();
             }
 
             sb.Append(i);
         }
 
-        if (sb is null) // FN
+        if (sb is null)                 // Noncompliant
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.CSharp8.cs:403]: This must be FP by now. sb is never null

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs [hunk 2]
```diff
@@ -169,18 +167,6 @@ public void Loops_For_ComplexCondition_MultipleVariables()
...
-                x[j].Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(9));
-                x[arg].Should().HaveOnlyConstraints(ObjectConstraint.NotNull, TestConstraint.First);
-            },
-            x =>
-            {
-                x[i].Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(1, 9));
-                x[j].Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(null, 0));
-                x[arg].Should().HaveOnlyConstraints(ObjectConstraint.NotNull, TestConstraint.First);
-            },
             x =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:170]: Why there are the final values that we have? For example the terminaction condition for `j=0` is
```
x[i].Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(2, 9));
x[j].Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(0));
```
I would expect `i` to be 0..9 here

Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:170]: Similar situation is on line 173. Why 0-8 and not 1-10?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/BranchingProcessor.cs [hunk 3]
```diff
@@ -41,7 +41,12 @@ protected override ProgramState[] Process(SymbolicContext context, T operation)
         var state = PreProcess(context.State, operation);
         if (BoolConstraintFromOperation(state, operation, context.IsLoopCondition, context.VisitCount) is { } constraint)
         {
-            return state.SetOperationConstraint(context.Operation, constraint).ToArray();    // We already know the answer from existing constraints
+            state = state.SetOperationConstraint(context.Operation, constraint);    // We already know the answer from existing constraints
+            if (context.VisitCount > 1)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/BranchingProcessor.cs:45]: What other processor except Binary benefits from this?

Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/BranchingProcessor.cs:45]: return ternary will be nicer (even with duplicated .ToArray())

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs [hunk 4]
```diff
@@ -272,12 +253,9 @@ public void Loops_While_FixedCount()
...
             x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(1)),
             x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(2, 10)));
-        validator.TagValues("After").Should().SatisfyRespectively(
-            x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(10, null)),
-            x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(10)));
-        validator.TagValues("End").Should().SatisfyRespectively(    // arg has only it's final constraints after looping once or twice
-            x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, TestConstraint.First),
-            x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, TestConstraint.First, BoolConstraint.True));
+        validator.TagValue("After").Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(10));
+        // arg has only it's final constraints after looping twice
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:257]: This typo is probably all over the file
```suggestion
        // arg has only its final constraints after looping twice
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
The complex double variable UT looks suspicious

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.CSharp8.cs (hunk 1 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.CSharp8.cs:403]: sb is never null => TP
or am I understanding you wrong?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 5]
```diff
@@ -201,8 +201,7 @@ private static SymbolicConstraint BinaryConstraint(BinaryOperatorKind kind, Symb
             return BinaryBoolConstraint(kind, leftBool == BoolConstraint.True, rightBool == BoolConstraint.True);
         }
         else if (left?.Constraint<NumberConstraint>() is { } leftNumber
-            && right?.Constraint<NumberConstraint>() is { } rightNumber
-            && EvaluateBranchingCondition(isLoopCondition, visitCount))
+            && right?.Constraint<NumberConstraint>() is { } rightNumber)
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:204]: That's coming in the next PR. Is it ok to keep it there or should I move it here?

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs (hunk 2 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:170]: When executing this code, the loop will iterate 10 times.
The SE engine will iterate it twice and determine that it cannot leave during these iterations. Afterward, it assumes that the loop will be left in a later iteration. In any case, the increment/decrement has happened at least twice.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/BranchingProcessor.cs (hunk 3 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/BranchingProcessor.cs:45]: I don't think any. Refactoring this logic into `Binary` will be very tricky, though.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/BranchingProcessor.cs [hunk 6]
```diff
@@ -41,7 +41,10 @@ protected override ProgramState[] Process(SymbolicContext context, T operation)
         var state = PreProcess(context.State, operation);
         if (BoolConstraintFromOperation(state, operation, context.IsLoopCondition, context.VisitCount) is { } constraint)
         {
-            return state.SetOperationConstraint(context.Operation, constraint).ToArray();    // We already know the answer from existing constraints
+            state = state.SetOperationConstraint(context.Operation, constraint);    // We already know the answer from existing constraints
+            return context.VisitCount > 1
```
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/BranchingProcessor.cs:45]: I would delete the comment above here - I'm not sure what it means with my current knowledge - and I would add a comment to line 45.
// We are branching for both conditions only for the second loop
of something similar.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs [hunk 7]
```diff
@@ -64,14 +64,10 @@ public void Loops_BodyVisitedMaxTwice_For_FixedCount(string loop)
...
-        validator.TagValues("InLoop").Should().SatisfyRespectively(
-            x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, TestConstraint.First),
-            x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, TestConstraint.First, BoolConstraint.True));
-        validator.TagValues("End").Should().SatisfyRespectively(    // Loop was entered, arg has only its final constraints
-            x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, TestConstraint.First),                        // looped once
-            x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, TestConstraint.First, BoolConstraint.True));  // looped twice
+        var validator = SETestContext.CreateCS(code, "int arg", new AddConstraintOnInvocationCheck()).Validator;
+        validator.ValidateExitReachCount(1);
+        // Loop was entered, arg has only its final constraints after looping twice
+        validator.TagValue("End").Should().HaveOnlyConstraints(ObjectConstraint.NotNull, TestConstraint.First, BoolConstraint.True);
```
Comment (mary-georgiou-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:70]: Is the `InLoop` tag seems to not be needed anymore here.

**Review by mary-georgiou-sonarsource — COMMENTED** [review]
Questiosn and nitpicks

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.CSharp8.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.CSharp8.cs:403]: Yes, you're right

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:204]: It's ok to split

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/BranchingProcessor.cs (hunk 6 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/BranchingProcessor.cs:45]: We do not branch here. We are just applying fuzzy logic to NumberConstraints.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs [hunk 8]
```diff
@@ -184,7 +179,7 @@ End If
             Next
             """;
         var validator = SETestContext.CreateVB(code, "Arg As Integer", new AddConstraintOnInvocationCheck()).Validator;
-        validator.ValidateExitReachCount(2);
+        validator.ValidateExitReachCount(1);
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:182]: Now only one branch leaves the loop/reaches the exit.

**Pull request merged**